File: agents/windows/src/N3NAgent.Core/Protocol/SecureMessage.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace N3NAgent.Core.Protocol;

/// <summary>
/// Secure message wrapper with encryption header.
/// Format: base64url(header).base64url(ciphertext).base64url(tag)
/// </summary>
public class SecureMessage
{
    public Header MessageHeader { get; set; } = new();
    public string Ciphertext { get; set; } = "";
    public string Tag { get; set; } = "";

    public class Header
    {
        [JsonPropertyName("v")]
        public int Version { get; set; } = 1;

        [JsonPropertyName("alg")]
        public string Algorithm { get; set; } = "A256GCM";

        [JsonPropertyName("did")]
        public string DeviceId { get; set; } = "";

        [JsonPropertyName("ts")]
        public long Timestamp { get; set; }

        [JsonPropertyName("seq")]
        public long Sequence { get; set; }

        [JsonPropertyName("nonce")]
        public string Nonce { get; set; } = "";

        [JsonPropertyName("dir")]
        public string Direction { get; set; } = "c2s";
    }

    /// <summary>
    /// Serialize to compact format: header.ciphertext.tag
    /// </summary>
    public string ToCompact()
    {
        var headerJson = JsonSerializer.Serialize(MessageHeader);
        var headerB64 = Base64UrlEncode(Encoding.UTF8.GetBytes(headerJson));
        return $"{headerB64}.{Ciphertext}.{Tag}";
    }

    /// <summary>
    /// Parse from compact format
    /// </summary>
    public static SecureMessage FromCompact(string compact)
    {
        var parts = compact.Split('.');
        if (parts.Length != 3)
            throw new SecureMessageException("Invalid secure message format");

        var headerJson = Encoding.UTF8.GetString(Base64UrlDecode(parts[0]));
        var header = JsonSerializer.Deserialize<Header>(headerJson)
            ?? throw new SecureMessageException("Invalid header");

        return new SecureMessage
        {
            MessageHeader = header,
            Ciphertext = parts[1],
            Tag = parts[2]
        };
    }

    /// <summary>
    /// Get the header as bytes for AAD
    /// </summary>
    public byte[] GetHeaderBytes()
    {
        return Encoding.UTF8.GetBytes(JsonSerializer.Serialize(MessageHeader));
    }

    public byte[] GetCiphertextBytes() => Base64UrlDecode(Ciphertext);
    public byte[] GetTagBytes() => Base64UrlDecode(Tag);
    public byte[] GetNonceBytes() => Base64UrlDecode(MessageHeader.Nonce);

    // Base64URL helpers
    public static string Base64UrlEncode(byte[] data)
    {
        return Convert.ToBase64String(data)
            .Replace('+', '-')
            .Replace('/', '_')
            .TrimEnd('=');
    }

    public static byte[] Base64UrlDecode(string data)
    {
        var base64 = data
            .Replace('-', '+')
            .Replace('_', '/');

        switch (base64.Length % 4)
        {
            case 2: base64 += "=="; break;
            case 3: base64 += "="; break;
        }

        return Convert.FromBase64String(base64);
    }
}

public class SecureMessageException : Exception
{
    public SecureMessageException(string message) : base(message) { }
}

File: src/main/java/com/aiinpocket/n3n/gateway/security/SecureMessage.java
package com.aiinpocket.n3n.gateway.security;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * Secure message wrapper with encryption header.
 * Format: base64url(header).base64url(ciphertext).base64url(tag)
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class SecureMessage {

    /**
     * Message header (transmitted in clear, but authenticated)
     */
    private Header header;

    /**
     * Encrypted payload
     */
    private String ciphertext;

    /**
     * GCM authentication tag
     */
    private String tag;

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Header {
        /**
         * Protocol version
         */
        @Builder.Default
        private int v = 1;

        /**
         * Encryption algorithm (always A256GCM)
         */
        @Builder.Default
        private String alg = "A256GCM";

        /**
         * Device ID
         */
        private String did;

        /**
         * Timestamp in milliseconds
         */
        private long ts;

        /**
         * Sequence number (increments with each message)
         */
        private long seq;

        /**
         * Nonce (Base64URL encoded)
         */
        private String nonce;

        /**
         * Direction: c2s (client to server) or s2c (server to client)
         */
        private String dir;
    }

    /**
     * Serialize to compact format: header.ciphertext.tag
     */
    public String toCompact(ObjectMapper objectMapper) {
        try {
            String headerJson = objectMapper.writeValueAsString(header);
            String headerB64 = Base64.getUrlEncoder().withoutPadding()
                .encodeToString(headerJson.getBytes(StandardCharsets.UTF_8));
            return headerB64 + "." + ciphertext + "." + tag;
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize secure message", e);
        }
    }

    /**
     * Parse from compact format
     */
    public static SecureMessage fromCompact(String compact, ObjectMapper objectMapper) {
        String[] parts = compact.split("\\.");
        if (parts.length != 3) {
            throw new IllegalArgumentException("Invalid secure message format: expected 3 parts");
        }

        try {
            String headerJson = new String(
                Base64.getUrlDecoder().decode(parts[0]),
                StandardCharsets.UTF_8
            );
            Header header = objectMapper.readValue(headerJson, Header.class);

            return SecureMessage.builder()
                .header(header)
                .ciphertext(parts[1])
                .tag(parts[2])
                .build();
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse secure message", e);
        }
    }

    /**
     * Get the header as bytes for AAD (Additional Authenticated Data)
     */
    @JsonIgnore
    public byte[] getHeaderBytes(ObjectMapper objectMapper) {
        try {
            return objectMapper.writeValueAsBytes(header);
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize header", e);
        }
    }

    /**
     * Get the ciphertext as bytes
     */
    @JsonIgnore
    public byte[] getCiphertextBytes() {
        return Base64.getUrlDecoder().decode(ciphertext);
    }

    /**
     * Get the tag as bytes
     */
    @JsonIgnore
    public byte[] getTagBytes() {
        return Base64.getUrlDecoder().decode(tag);
    }

    /**
     * Get the nonce as bytes
     */
    @JsonIgnore
    public byte[] getNonceBytes() {
        return Base64.getUrlDecoder().decode(header.getNonce());
    }
}
